Civi/Api4/Service/Links/MembershipLinksProvider.php

```diff
@@ -0,0 +1,133 @@
+        if (!empty($paymentObject)) {
+          $isUpdateBilling = $paymentObject->supports('updateSubscriptionBillingInfo');
+        }
+        if (!$isUpdateBilling) {
+          self::unsetLinks($links, ['changebilling']);
+        }
+        $isCancelSupported = \CRM_Member_BAO_Membership::isCancelSubscriptionSupported($membershipId);
+        if (!$isCancelSupported) {
+          self::unsetLinks($links, ['cancelrecur']);
+        }
+      }
+
+      // Unset renew and followup for deceased memberships.
+      $membershipStatus = $request->getValue('status_id:name');
+      if ($membershipStatus && $membershipStatus === 'Deceased') {
+        self::unsetLinks($links, ['renew', 'followup']);
+      }
+
+      $membershipTypeId = $request->getValue('membership_type_id');
+      if ($membershipTypeId && \CRM_Financial_BAO_FinancialType::isACLFinancialTypeStatus()) {
+        $finType = \Civi\Api4\MembershipType::get(TRUE)
+          ->addSelect('financial_type_id:name')
+          ->addWhere('id', '=', $membershipTypeId)
+          ->execute()
+          ->first()['financial_type_id:name'] ?? NULL;
+        if ($finType && !\CRM_Core_Permission::check('edit contributions of type ' . $finType)) {
+          self::unsetLinks($links, ['update', 'renew', 'followup']);
+        }
+        if ($finType && !\CRM_Core_Permission::check('delete contributions of type ' . $finType)) {
+          self::unsetLinks($links, ['delete']);
+        }
+      }
+
+      $e->getResponse()->exchangeArray(array_values($links));
+    }
+  }
+
+  private static function addLinks(array &$newLinks, array $addTemplate) {
+    $actions = [
+      'renew' => [
+        'path' => 'civicrm/contact/view/membership?action=renew&reset=1&cid=[contact_id]&id=[id]&context=membership&selectedChild=member',
+        'text' => ts('Renew Membership'),
+      ],
+      'followup' => [
+        'path' => 'civicrm/contact/view/membership?action=renew&reset=1&cid=[contact_id]&id=[id]&context=membership&selectedChild=member&mode=live',
+        'text' => ts('Renew-Credit Card Membership'),
+      ],
+      'cancelrecur' => [
+        'path' => 'civicrm/contribute/unsubscribe?reset=1&cid=[contact_id]&mid=[id]&context=membership&selectedChild=member',
+        'text' => ts('Cancel Auto-renewal'),
+      ],
+      'changebilling' => [
+        'path' => 'civicrm/contribute/updatebilling?reset=1&cid=[contact_id]&mid=[id]&context=membership&selectedChild=member',
+        'text' => ts('Change Billing Details'),
+      ],
+    ];
+    foreach ($actions as $action => $values) {
+      $addTemplate['ui_action'] = $action;
+      $addTemplate['path'] = $values['path'];
+      $addTemplate['text'] = $values['text'];
+      $newLinks[] = $addTemplate;
+    }
+  }
+
+  private static function unsetLinks(array &$links, array $actions) {
+    foreach ($actions as $action) {
+      $actionLinkIndex = self::getActionIndex($links, $action);
+      unset($links[$actionLinkIndex]);
```

Comment:
Probably need to wrap this `unset` in a conditional like `if (isset($actionLinkIndex))` since `getActionIndex()` could return `null` (this could happen because of filtering by the WHERE clause excluding the links so they don't need to be unset.).